python/cudf_polars/cudf_polars/experimental/scheduler.py

```diff
@@ -0,0 +1,171 @@
+    return [
+        k
+        for k in (
+            maybe_task[1:]
+            if istask(maybe_task)
+            else [maybe_task]  # maybe_task might be a key
+        )
+        if is_hashable(k) and k in graph
+    ]
+
+
+def toposort(graph: Graph, dependencies: Mapping[Key, list[Key]]) -> list[Key]:
+    """Return a list of task keys sorted in topological order."""
+    # Stack-based depth-first search traversal. This is based on Tarjan's
+    # method for topological sorting (see wikipedia for pseudocode)
+    ordered: list[Key] = []
+    completed: set[Key] = set()
+    seen: set[Key] = set()
+
+    for key in graph:
+        if key in completed:
+            continue
+        nodes: list[Key] = [key]
+        while nodes:
+            # Keep current node on the stack until all descendants are visited
+            current = nodes[-1]
+            if current in completed:  # pragma: no cover
+                # Already fully traversed descendants of current
+                nodes.pop()
+                continue
+            seen.add(current)
+
+            # Add direct descendants of current to nodes stack
+            next_nodes = []
+            for nxt in dependencies[current]:
+                if nxt not in completed:
+                    if nxt in seen:  # pragma: no cover
+                        # Cycle detected!
+                        raise RuntimeError(
+                            f"Cycle detected in the task graph! Key: {nxt}"
+                        )
+                    next_nodes.append(nxt)
+
+            if next_nodes:
+                nodes.extend(next_nodes)
+            else:
+                # Current has no more descendants to explore
+                ordered.append(current)
+                completed.add(current)
+                seen.remove(current)
+                nodes.pop()
+
+    return ordered
+
+
+def synchronous_scheduler(
+    graph: Graph,
+    key: Key,
+    *,
+    cache: MutableMapping | None = None,
+) -> Any:
+    """
+    Execute the task graph for a given key.
+
+    Parameters
+    ----------
+    graph
+        The task graph to execute.
+    key
+        The final output key to extract from the graph.
+    cache
+        Intermediate-data cache.
+
+    Returns
+    -------
+    Executed task-graph result for ``key``.
+    """
+    if key not in graph:  # pragma: no cover
+        raise KeyError(f"{key} is not a key in the graph")
+    if cache is None:
+        cache = {}
+
+    refcount: defaultdict[Key, int] = defaultdict(int)
+    dependencies: dict[Key, list] = {}
+    for k in graph:
+        dependencies[k] = required_keys(k, graph)
+        for dep in dependencies[k]:
+            refcount[dep] += 1
+
+    for k in toposort(graph, dependencies):
+        task = graph[k]
+        result = _execute_task(task, cache)
+        cache[k] = result
+        for dep in dependencies[k]:
+            refcount[dep] -= 1
+            if refcount[dep] == 0 and dep != key:
```

Comment:
Then here we only need to check `refcount[dep] == 0`.

Reply:
The `dep != key` check is already unnecessary if the graph is generated "correctly". Note that this is checking that `dep != key`, not that `k != key`.